plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -819,6 +855,106 @@ private void finishOptimize(ConnectorSession session, IcebergTableExecuteHandle
         transaction = null;
     }
 
+    @Override
+    public void executeTableExecute(ConnectorSession session, ConnectorTableExecuteHandle tableExecuteHandle)
+    {
+        IcebergTableExecuteHandle executeHandle = (IcebergTableExecuteHandle) tableExecuteHandle;
+        switch (executeHandle.getProcedureId()) {
+            case VACUUM:
+                executeVacuum(session, executeHandle);
+                return;
+            default:
+                throw new IllegalArgumentException("Unknown procedure '" + executeHandle.getProcedureId() + "'");
+        }
+    }
+
+    private void executeVacuum(ConnectorSession session, IcebergTableExecuteHandle executeHandle)
+    {
+        IcebergVacuumHandle vacuumHandle = (IcebergVacuumHandle) executeHandle.getProcedureHandle();
+
+        Table table = catalog.loadTable(session, executeHandle.getSchemaTableName());
+        int tableVersion = ((BaseTable) table).operations().current().formatVersion();
+        if(tableVersion > VACUUM_MAX_SUPPORTED_TABLE_VERSION) {
+            throw new TrinoException(NOT_SUPPORTED, format("VACUUM is not supported for iceberg table version > %d. " +
+                    "Table %s version is %s.",
+                    VACUUM_MAX_SUPPORTED_TABLE_VERSION,
+                    executeHandle.getSchemaTableName(),
+                    tableVersion));
+        }
+        Map<String, String> properties = table.properties();
+        if (properties.containsKey(WRITE_LOCATION_PROVIDER_IMPL)) {
+            throw new TrinoException(NOT_SUPPORTED, "Table " + executeHandle.getSchemaTableName() + " specifies " + properties.get(WRITE_LOCATION_PROVIDER_IMPL) +
+                    " as a location provider. Writing to Iceberg tables with custom location provider is not supported.");
+        }
+        Duration retention = requireNonNull(vacuumHandle.getRetentionThreshold(), "retention is null");
+        Duration minRetention = getVacuumMinRetention(session);
+        checkProcedureArgument(retention.compareTo(minRetention) >= 0,
+                "Retention specified (%s) is shorter than the minimum retention configured in the system (%s). " +
+                        "Minimum retention can be changed with %s configuration property or iceberg.%s session property",
+                retention,
+                minRetention,
+                IcebergConfig.VACUUM_MIN_RETENTION,
+                IcebergSessionProperties.VACUUM_MIN_RETENTION);
+
+        long expireTimestampMillis = session.getStart().toEpochMilli() - retention.toMillis();
+        table.expireSnapshots()
```

Comment:
We should make sure this is well documented. As is, `vacuum` does not just remove orphaned files, but it also expires snapshots, removing history from the table.

There may be an argument for having that be two different procedures, `ExpireSnapshots` and `Vacuum`

Reply:
Maybe, we may discuss that. tough it should be done before implementing not after :D